File: P1S2_Ruby/docencia.rb
require_relative 'aula_practicas'

class Docencia
  def initialize(factoria_docencia)
    @factoria_docencia = factoria_docencia
    @aulas = []
    @grupos = []
  end

  def add_aula(nombre_aula)
    @aulas << @factoria_docencia.crear_aula(nombre_aula)
  end

  def add_grupo(nombre_grupo)
    @grupos << @factoria_docencia.crear_grupo(nombre_grupo)
  end

  def asignar_aula_grupo(nombre_aula, nombre_grupo)
    grupo_buscado = @grupos.find { |item| item.nombre == nombre_grupo }
    raise ArgumentError, 'Grupo solicitado no existe' if grupo_buscado.nil?

    aula_buscada = @aulas.find { |item| item.nombre == nombre_aula }
    raise ArgumentError, 'Aula solicitada no existe' if aula_buscada.nil?

    grupo_buscado.asignar_aula(aula_buscada)
  end

  def add_ordenador(nombre_aula, ordenador)
    aula_buscada = @aulas.find { |item| item.nombre == nombre_aula }
    raise ArgumentError, 'Aula solicitada no existe' if aula_buscada.nil?

    raise ArgumentError, 'Aula solicitada no es de prácticas' unless aula_buscada.is_a? AulaPracticas

    aula_buscada.add_ordenador(ordenador)
  end

  def to_s
    s = "Docencia{grupos=["
    @grupos.each { |grupo|
      s += grupo.to_s
    }
    s += "], aulas=#{@aulas}}"
  end
end

File: P1S2_Java/src/Docencia.java
import java.util.ArrayList;
import java.util.NoSuchElementException;

public class Docencia {
    FactoriaDocencia factoriaDocencia;
    ArrayList<Aula> aulas = new ArrayList<>();
    ArrayList<Grupo> grupos = new ArrayList<>();

    public Docencia(FactoriaDocencia factoriaDocencia) {
        this.factoriaDocencia = factoriaDocencia;
    }

    public void addAula(String nombreAula) {
        aulas.add(factoriaDocencia.crearAula(nombreAula));
    }

    public void addGrupo(String nombreGrupo) {
        grupos.add(factoriaDocencia.crearGrupo(nombreGrupo));
    }

    public void asignarAulaGrupo(String nombreAula, String nombreGrupo) throws NoSuchElementException {
        Grupo grupoBuscado = buscarGrupo(nombreGrupo);
        if(grupoBuscado == null)
            throw new NoSuchElementException("No se ha encontrado el grupo");

        Aula aulaBuscada = buscarAula(nombreAula);
        if(aulaBuscada == null)
            throw new NoSuchElementException("No se ha encontrado el aula");

        grupoBuscado.asignarAula(aulaBuscada);
    }

    private Aula buscarAula(String nombreAula) {
        return aulas.stream()
                .filter(elemento -> nombreAula.equals(elemento.getNombre()))
                .findAny()
                .orElse(null);
    }

    private Grupo buscarGrupo(String nombreGrupo) {
        return grupos.stream()
                .filter(elemento -> nombreGrupo.equals(elemento.getNombre()))
                .findAny()
                .orElse(null);
    }

    @Override
    public String toString() {
        final StringBuilder sb = new StringBuilder("Docencia{");
        //sb.append("aulas=").append(aulas);
        sb.append("grupos=").append(grupos);
        sb.append('}');
        return sb.toString();
    }

    public void addOrdenador(String nombreAula, Ordenador ordenador) throws NoSuchElementException {
        AulaPracticas aulaBuscada = (AulaPracticas)buscarAula(nombreAula);
        if(aulaBuscada == null)
            throw new NoSuchElementException("No se ha encontrado el aula");

        aulaBuscada.addOrdenador(ordenador);
    }
}
